keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,793 @@
+   - "publicReadOnly": Read only, uncredentialed users can call ListBucket and GetObject.
+   - "publicWriteOnly": Write only, uncredentialed users can only call PutObject.
+   - "publicReadWrite": Read/Write, same as `ro` with the addition of PutObject being allowed.
+> Note: does not reflect or alter the backing storage ACLs or IAM policies.
+7. `bucketClassName`: Name of the associated bucket class.
+8. `bucketRequest`: Name and namespace of the associated `BucketRequest`.
+9. `allowedNamespaces`: a list of namespaces that are permitted to either create new buckets or to access existing buckets.
+10. `protocol`: The protocol the application will use to access the backend storage.
+   - `protocolSignature`: Specifies the protocol targeted by this Bucket instance.  One of:
+     - `azureBlob`: data required to target a provisioned azure container and/or storage account.
+     - `s3`: data required to target a provisioned S3 bucket and/or user.
+     - `gcs`: data required to target a provisioned GCS bucket and/or service account.
+11. `parameters`: a copy of the BucketClass parameters.
+12. `bucketAvailable`: if true the bucket has been provisioned. If false then the bucket has not been provisioned and is unable to be accessed.
+
+#### BucketClass
+
+A cluster-scoped custom resource to provide admins control over the handling of bucket provisioning.  The `BucketClass` (BC) defines a retention policy, driver specific parameters, and the provisioner name. A list of allowed namespaces can be specified to restrict new bucket creation and access to existing buckets. A default bucket class can be defined for each supported protocol. This allows the bucket class to be omitted from a `BucketRequest`. All of the `BucketClass` fields are copied to the `Bucket` instance.
+
+> Note: like storage classes, a `BucketClass` is immutable, but a `Bucket` is mutable and contains all of the BC fields.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketClass
+metadata:
+  name: 
+provisioner: [1]
+isDefaultBucketClass: [2]
+protocol: {"azureblob", "gcs", "s3", ... } [3]
+anonymousAccessMode: [4]
+retentionPolicy: {"Delete", "Retain"} [5]
+allowedNamespaces: [6]
+  - name:
+parameters: [7]
+```
+
+1. `provisioner`: the name of the driver. If supplied the driver container and sidecar container are expected to be deployed. Format: <provisioner-namespace>"/"<provisioner-name>, eg "ceph-rgw-provisoning/ceph-rgw.cosi.ceph.com".
+2. `isDefaultBucketClass`: (optional) boolean, default is false. If set to true then a `BucketRequest` may not need to specify a `BucketClass`. If the greenfield `BucketRequest` omits the `BucketClass` and a default `BucketClass`'s protocol matches the `BucketRequest`'s protocol then the default bucket class is used; otherwise an error is logged.
+3. `protocol`: (required) protocol supported by the associated object store. This field validates that the `BucketRequest`'s desired protocol is supported.
```

Comment:
Is this a fixed enum including just s3/blob/gcs? 
or is this a free form string that COSI just matches between the BR and the BC?
I mean - can a provider decide to support more variants like "s3:v2" or "s3:ceph"?

Reply:
Good question! It is proposed to be a simple string match, but version has come up in reviews and has not been reflected in the KEP other than in B.protocol.s3.vesion. I think we need to support something like you've described. 
@wlan0 @brahmaroutu

Reply:
Right, we discussed at least the S3 apiVersion -
https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/locking-api-versions.html